File: src/test/kotlin/com/yosep/msa/yoscouponapi/coupon/service/CouponServiceTest.java
//package com.yosep.msa.yoscouponapi.coupon.service;
//
//import com.yosep.msa.yoscouponapi.coupon.domain.Coupon;
//import com.yosep.msa.yoscouponapi.coupon.domain.CouponDTO;
//import com.yosep.msa.yoscouponapi.coupon.repository.CouponRepository;
//import com.yosep.msa.yoscouponapi.stock.domain.Stock;
//import org.junit.Assert;
//import org.junit.jupiter.api.Test;
//import org.junit.runner.RunWith;
//import org.springframework.beans.factory.annotation.Autowired;
//import org.springframework.boot.test.context.SpringBootTest;
//import org.springframework.data.redis.core.ReactiveRedisTemplate;
//import org.springframework.data.redis.core.RedisTemplate;
//import org.springframework.test.annotation.Rollback;
//import org.springframework.test.context.ActiveProfiles;
//import org.springframework.test.context.junit4.SpringRunner;
//
//import javax.annotation.Resource;
//import java.util.HashSet;
//import java.util.Set;
//
//import static org.hamcrest.CoreMatchers.is;
//
//@RunWith(SpringRunner.class)
//@SpringBootTest
//public class CouponServiceTest {
//    @Resource
//    private RedisTemplate<String, Coupon> redisTemplate;
//
//    @Autowired
//    private ReactiveRedisTemplate<String, String> reactiveRedisTemplate;
//
//    @Autowired
//    CouponRepository couponRepository;
//
//    @Test
//    @Rollback(value = true)
//    public void createTest() {
//        // Given
//        CouponDTO couponDTO = new CouponDTO("coupon_test_1","쿠폰테스트1",10);
//        Coupon coupon = Coupon.Companion.of(couponDTO);
//
//        Stock stock = Stock.Companion.of(couponDTO,coupon);
//        coupon.setStock(stock);
//
//        // When
//        Coupon createdCoupon = couponRepository.save(coupon);
//
//        // Then
//        Assert.assertThat(createdCoupon.getCouponId(),is("coupon_test_1"));
//        Assert.assertThat(createdCoupon.getName(),is("쿠폰테스트1"));
//        Assert.assertThat(createdCoupon.getStock().getTotal(),is(10));
//    }
//
//    @Test
//    public void opsValue() {
//        reactiveRedisTemplate.opsForValue();
//        Set<String> cacheKeys = new HashSet<>();
//
////        reactiveRedisTemplate.
//    }
//
//}

File: src/test/kotlin/com/yosep/msa/yoscouponapi/coupon/service/CouponServiceTest.kt
package com.yosep.msa.yoscouponapi.coupon.service

import com.yosep.msa.yoscouponapi.coupon.domain.CouponForUser
import com.yosep.msa.yoscouponapi.coupon.domain.CouponState
import com.yosep.msa.yoscouponapi.coupon.repository.CouponRepository
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.DisplayNameGeneration
import org.junit.jupiter.api.DisplayNameGenerator
import org.junit.jupiter.api.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.annotation.Rollback
import org.springframework.test.context.junit4.SpringRunner
import java.time.LocalDateTime
import java.util.*

@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores::class)
@RunWith(SpringRunner::class)
@Rollback(value= true)
@SpringBootTest
class CouponServiceTest(
        @Autowired
        var couponRepository: CouponRepository
) {
    @Test
    @DisplayName(value="coupon 생성 repository 테스트")
    fun createCouponTest() {

        var coupon = CouponForUser(
                "coupon-create-test",
                "coupon-create-test",
                CouponState.START,
                1,
                LocalDateTime.now(),
                LocalDateTime.now())

        var createdCoupon = couponRepository.save(coupon)
        println(createdCoupon.toString())

        var couponId:String = "coupon-create-test"
        var findedCoupon = couponRepository.findById(couponId)

        while (!findedCoupon.isEmpty) {
            println("중복")
            break
        }

        do {
            couponId = "coupon_" + UUID.randomUUID()
            var findedCoupon = couponRepository.findById(couponId)
        }while (!findedCoupon.isEmpty)

        var newCoupon = CouponForUser(
                couponId,
                "coupon-create-test",
                CouponState.START,
                1,
                LocalDateTime.now(),
                LocalDateTime.now()
        )

        var newCreatedCoupon = couponRepository.save(newCoupon)

        println(newCreatedCoupon.toString())
    }
}